enterprise/server/remote_execution/snaploader/snaploader.go

```diff
@@ -42,21 +43,58 @@ const (
 	chunkedFileWriteConcurrency = 4
 )
 
-// NewKey returns the cache key for a snapshot.
+// SnapshotKeySet returns the cache keys for potential snapshot matches,
+// as well as the key that should be written to.
+//
 // TODO: include a version number in the key somehow, so that
 // if we make breaking changes e.g. to the vmexec API or firecracker
 // version etc., we can ensure that incompatible snapshots don't get reused.
-func NewKey(task *repb.ExecutionTask, configurationHash, runnerID string) (*fcpb.SnapshotKey, error) {
+func SnapshotKeySet(task *repb.ExecutionTask, configurationHash, runnerID string) (*fcpb.SnapshotKeySet, error) {
 	pd, err := digest.ComputeForMessage(task.GetCommand().GetPlatform(), repb.DigestFunction_SHA256)
 	if err != nil {
 		return nil, status.WrapErrorf(err, "failed to compute platform hash")
 	}
-	return &fcpb.SnapshotKey{
-		InstanceName:      task.GetExecuteRequest().GetInstanceName(),
-		PlatformHash:      pd.GetHash(),
-		ConfigurationHash: configurationHash,
-		RunnerId:          runnerID,
-	}, nil
+
+	branchRef, fallbackRefs := gitRefs(task)
+	keys := &fcpb.SnapshotKeySet{
+		BranchKey: &fcpb.SnapshotKey{
+			InstanceName:      task.GetExecuteRequest().GetInstanceName(),
+			PlatformHash:      pd.GetHash(),
+			ConfigurationHash: configurationHash,
+			Ref:               branchRef,
+			RunnerId:          runnerID,
+		},
+	}
+	for _, ref := range fallbackRefs {
+		fallbackKey := proto.Clone(keys.BranchKey).(*fcpb.SnapshotKey)
+		fallbackKey.Ref = ref
+		keys.FallbackKeys = append(keys.FallbackKeys, fallbackKey)
+	}
+	return keys, nil
+}
+
+func gitRefs(task *repb.ExecutionTask) (branchRef string, fallbackRefs []string) {
+	// NOTE: keep these names in sync with workflow service
+	branchRef = getEnv(task, "GIT_BRANCH")
+	if branchRef == "" {
+		return "", nil
```

Comment:
What does this case mean? Should we either return an error that the env vars weren't set correctly, or use the fallbacks if they're set?

Reply:
I figured if `GIT_BRANCH` is not set, then we're not looking at a well-formed workflow action, so to be on the safe side we should ignore the fallbacks. Added a clarifying comment.

Reply:
This scares me a little -- are only workflows is allowed to use snapshotting? What about recyclable runners with container-isolation-type=firecracker?

Reply:
Those will still be allowed to use snapshotting but just won't have a git ref as part of the snapshot key.

Reply:
Is there a reason we don't want to include git ref for these non-workflow tasks? Seems like it would improve their snapshot match quality

Reply:
In general, actions will not have these git ref env vars set. We are setting these specially for workflow actions.

Reply:
It's unclear to me whether git ref is the right thing there, vs something like a hash of the Action proto.